drivers/adc/adc_esp32.c

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright (c) 2022 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_adc
+
+#include <errno.h>
+#include <hal/adc_hal.h>
+#include <hal/adc_types.h>
+#include <esp_adc_cal.h>
+#include <esp_heap_caps.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/adc.h>
+#include "driver/periph_ctrl.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(adc_esp32, CONFIG_ADC_LOG_LEVEL);
+
+#if CONFIG_IDF_TARGET_ESP32
+#define ADC_CALI_SCHEME		ESP_ADC_CAL_VAL_EFUSE_VREF
+#define ADC_RESOLUTION_MIN	SOC_ADC_DIGI_MIN_BITWIDTH
+#define ADC_RESOLUTION_MAX	SOC_ADC_DIGI_MAX_BITWIDTH
+
+/* Due to significant measurement discrepancy in higher voltage range, we
+ * clip the value instead of yet another correction. The IDF implementation
+ * for ESP32-S2 is doing it, so we copy that approach in Zephyr driver
+ */
+#define ADC_CLIP_MVOLT_11DB	2550
+
+#elif CONFIG_IDF_TARGET_ESP32S2
+#define ADC_CALI_SCHEME		ESP_ADC_CAL_VAL_EFUSE_TP
+#define ADC_RESOLUTION_MIN	SOC_ADC_DIGI_MAX_BITWIDTH
+#define ADC_RESOLUTION_MAX	SOC_ADC_MAX_BITWIDTH
+
+#elif CONFIG_IDF_TARGET_ESP32C3
+#define ADC_CALI_SCHEME		ESP_ADC_CAL_VAL_EFUSE_TP
+#define ADC_RESOLUTION_MIN	SOC_ADC_DIGI_MAX_BITWIDTH
+#define ADC_RESOLUTION_MAX	SOC_ADC_DIGI_MAX_BITWIDTH
+
+#endif
+
+/* Convert resolution in bits to esp32 enum values */
+#define WIDTH_MASK(r) ((((r) - 9) < ADC_WIDTH_MAX) ? ((r) - 9) : (ADC_WIDTH_MAX - 1))
+
+/* Validate if resolution in bits is within allowed values */
+#define VALID_RESOLUTION(r) ((r) >= ADC_RESOLUTION_MIN && (r) <= ADC_RESOLUTION_MAX)
+#define INVALID_RESOLUTION(r) (!VALID_RESOLUTION(r))
+
+/* Default internal reference voltage */
+#define ADC_ESP32_DEFAULT_VREF_INTERNAL (1100)
+
+struct adc_esp32_conf {
+	adc_unit_t unit;
+	uint8_t channel_count;
+};
+
+struct adc_esp32_data {
+	adc_atten_t attenuation[ADC_CHANNEL_MAX];
+	uint8_t resolution[ADC_CHANNEL_MAX];
+	esp_adc_cal_characteristics_t chars[ADC_CHANNEL_MAX];
+	uint16_t meas_ref_internal;
+	uint16_t *buffer;
+	uint16_t *buffer_repeat;
+	bool calibrate;
+};
+
+/* Convert zephyr,gain property to the ESP32 attenuation */
+static inline int gain2attenuation(enum adc_gain gain, adc_atten_t *atten)
+{
+	switch (gain) {
+	case ADC_GAIN_1:
+		*atten = ADC_ATTEN_DB_0;
+		break;
+	case ADC_GAIN_4_5:
+		*atten = ADC_ATTEN_DB_2_5;
+		break;
+	case ADC_GAIN_1_2:
+		*atten = ADC_ATTEN_DB_6;
+		break;
+	case ADC_GAIN_1_4:
+		*atten = ADC_ATTEN_DB_11;
+		break;
+	default:
+		return -1;
+	}
+	return 0;
+}
+
+/* Convert voltage by inverted attenuation to support zephyr gain values */
+static void attenuation2gain_invert(adc_atten_t atten, uint32_t *val_mv)
+{
+	if (!val_mv) {
+		return;
```

Comment:
How this error is handled in upper layers, since null here implies in a noop?

Reply:
in case no `val_mv` is provided nothing would converted making result incorrect. but this is just helper static function not exposed elsewhere, and the check is probably not needed.